### PR TITLE
partial bitcoin#15638, #21966, #16889, merge #14555, #20499, #14074, #17073: util refactoring

### PR DESCRIPTION
**Pull request by kittywhiskers**
## Contents

bitcoin#14074, bitcoin#14555, bitcoin#17073, bitcoin#20499

### Partials

bitcoin#15638, bitcoin#21966, bitcoin#16889

## Notes

_With scripted diffs with the command used available in the commit message, drop the `g` from `gsed` if you're not running macOS. This is a work in progress, more scripted refactoring to follow._

### CODE REVIEW

File: src/util/error.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2010-2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/error.h>
+
+#include <util/system.h>
+
+std::string TransactionErrorString(const TransactionError err)
```
Comment (PastaPastaPasta) [inline src/util/error.cpp:9]: This commit should be "moving" this, but we don't have it yet. Should maybe backport what adds this?

Comment (kittywhiskers) [reply src/util/error.cpp:9]: Introduced in a series of changes brought in by https://github.com/bitcoin/bitcoin/pull/14978, cannot be merged as PSBT is not merged. Will remove line

File: src/rpc/mining.cpp [hunk 2]
```diff
@@ -22,8 +22,11 @@
 #include <rpc/mining.h>
 #include <rpc/server.h>
 #include <txmempool.h>
-#include <util.h>
-#include <utilstrencodings.h>
+#include <util/fees.h>
+#include <util/strencodings.h>
+#include <util/system.h>
+#include <util/validation.h>
+#include <validation.h>
```
Comment (UdjinM6) [inline src/rpc/mining.cpp:29]:
```
Duplicate include(s) in src/rpc/mining.cpp:
#include <validation.h>
```
https://gitlab.com/dashpay/dash/-/jobs/1375020079#L55
should drop this line or (even better) a similar include at line 15

File: src/validation.cpp [hunk 3]
```diff
@@ -34,9 +34,11 @@
 #include <txmempool.h>
 #include <ui_interface.h>
 #include <undo.h>
-#include <util.h>
-#include <utilmoneystr.h>
-#include <utilstrencodings.h>
+#include <util/system.h>
```
Comment (UdjinM6) [inline src/validation.cpp:37]:
```
Duplicate include(s) in src/validation.cpp:
#include <util/system.h>
```
https://gitlab.com/dashpay/dash/-/jobs/1375020079#L57

File: src/consensus/tx_verify.cpp [hunk 4]
```diff
@@ -13,6 +13,7 @@
 #include <chain.h>
 #include <coins.h>
 #include <util/moneystr.h>
+#include <version.h>
```
Comment (PastaPastaPasta) [inline src/consensus/tx_verify.cpp:16]: why was this added (15638)?

File: src/init.cpp [hunk 5]
```diff
@@ -44,9 +44,11 @@
 #include <txmempool.h>
 #include <torcontrol.h>
 #include <ui_interface.h>
-#include <util/system.h>
 #include <util/asmap.h>
+#include <util/error.h>
 #include <util/moneystr.h>
+#include <util/system.h>
+#include <util/validation.h>
 #include <validationinterface.h>
```
Comment (PastaPastaPasta) [inline src/init.cpp:52]: quite different to upstream

File: src/consensus/tx_verify.cpp (hunk 4 above)
Comment (kittywhiskers) [reply src/consensus/tx_verify.cpp:16]: To fix compile errors

File: src/init.cpp (hunk 5 above)
Comment (kittywhiskers) [reply src/init.cpp:52]: To fix compile errors

File: src/hash.h [hunk 6]
```diff
@@ -264,27 +264,6 @@ unsigned int MurmurHash3(unsigned int nHashSeed, Span<const unsigned char> vData
...
-     *  This function can only be used when a multiple of 8 bytes have been written so far.
-     */
-    CSipHasher& Write(uint64_t data);
-    /** Hash arbitrary bytes. */
-    CSipHasher& Write(const unsigned char* data, size_t size);
-    /** Compute the 64-bit SipHash-2-4 of the data written so far. The object remains untouched. */
-    uint64_t Finalize() const;
-};
 
 /** Optimized SipHash-2-4 implementation for uint256.
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/hash.h:268]: Resolved

File: src/test/util_tests.cpp [hunk 7]
```diff
@@ -11,6 +11,7 @@
 #include <util/string.h>
```
[the remark this answers is not in the document]
Comment (kittywhiskers) [reply src/test/util_tests.cpp:11]: Will add #13072 to backports list, mark #16889 as partial

**Comment by PastaPastaPasta** [comment]
Seems like needs a rebase, otherwise force pushes look as expected

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Review by UdjinM6 — APPROVED** [review]
utACK
